### PR TITLE
fix(#1372): Set Kind/Filename mapping as AsciiDoc file

### PR DESCRIPTION
**Pull request by lordofthejars**
The mapping between Kind and Filename is moved to an AsciiDoc file, so it can be reused in documentation as well.

By default, it loads this mapping from `/fabric8/default-kind-filename-mapping.adoc` classpath location, but using env variable or system property named `fabric8.mapping` you can specify your own file.

Initially will inspect classpath and if it returns null, then it will treat `fabric8.mapping` value as a disk location.

### CODE REVIEW

**Comment by lordofthejars** [comment]
Yes, I'll add the line in CHANGELOG.

Regarding this, I think we could create some method that automatically updates the `CHANGELOG` and about licenses, I have found some files with and without the headers, but at the same time, this could be also automatized. 

So for now, I will do it manually but it will be great to think about some kind of automation of the PR.

**Comment by rhuss** [comment]
We could consider using [commitizen](https://commitizen.github.io/cz-cli/) for creating the [changelogs](https://github.com/conventional-changelog/conventional-changelog), but it requires quite some discipline when doing commit messages (and also every PR needs to be squashed). 

We tried that for Syndesis, but ended up to manage the changelogs manually again, because at the end it doesn't matter whether you have to remind people to add to the Changelog or whether to remind them to change the commit messages. Manually editing the Changelog also give you a better chance for rewording, as the changelog is more target to an end user than to a developer (like commit messages do).

tbh, I wouldn't bother yet with auto generating of changelogs yet.

Checking license files is something different and we should add the license plugin again.

File: core/src/main/resources/fabric8/default-kind-filename-mapping.adoc [hunk 1]
```diff
@@ -0,0 +1,138 @@
+|===
+|Kind |Filename
```
Comment (rhuss) [inline core/src/main/resources/fabric8/default-kind-filename-mapping.adoc]: The first column is ist the filename part indicating the kind, the second then is the kind.

Maybe `Filename` is also a bit misleading as one could think about this as a whole file name but its only the last part before the extension. Hard to find a good name here, as the extension is obviously ".yml", so it something like a 'file indicator' or so. wdyt ?

File: core/src/test/java/io/fabric8/maven/core/util/AsciiDocParserTest.java [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+        + "|cm" + System.lineSeparator()
+        + "a|ConfigMap" + System.lineSeparator()
+        + System.lineSeparator()
+        + "|cronjob" + System.lineSeparator()
+        + "|CronJob" + System.lineSeparator()
+        + "|===";
+
+    private static final String NONE_END_VALID_TABLE = "cols=2*,options=\"header\"]" + System.lineSeparator()
+        + "|===" + System.lineSeparator()
+        + "|Kind" + System.lineSeparator()
```
Comment (rhuss) [inline core/src/test/java/io/fabric8/maven/core/util/AsciiDocParserTest.java]: For consistencies sake, I would change this header, too (also it doesnt really matter in the test)

File: core/src/main/resources/fabric8/default-kind-filename-mapping.adoc [hunk 3]
```diff
@@ -0,0 +1,138 @@
+|===
```
Comment (rhuss) [inline core/src/main/resources/fabric8/default-kind-filename-mapping.adoc]: As this file will be updated when a new resource is to be added, we should add an asciidoc comment here to explain what can be update how.

* don't add extra newlines before the new column
* E.g. no multiline columns are allowed (each column must be an extra line)
* Keep Kind, filenames in a sorted order

etc.

File: core/src/main/resources/fabric8/default-kind-filename-mapping.adoc (hunk 1 above)
Comment (rhuss) [reply core/src/main/resources/fabric8/default-kind-filename-mapping.adoc]: Tbh, I wouldn't change the header but the order of the columns because I expect users to look up which file indicator is used for which Kind, so look up the Kind (e.g. 'ConfigMap') to find out what to use in the fragment file (`-cm.yml`). As the table is supposed to be ordered alphabetically in the first column, it would be awesome if the kind would be in the first column.

I know this required some code and test changes, but I believe it would be worth before go out in the wild with it.

File: core/src/main/java/io/fabric8/maven/core/util/kubernetes/KindFilenameMapperUtil.java [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.io.InputStream;
+
+public class KindFilenameMapperUtil {
+
+    public static String[] loadMappings() {
+
+        final String location =
+            EnvUtil.getEnvVarOrSystemProperty("fabric8.mapping", "/fabric8/default-kind-filename-mapping.adoc");
```
Comment (rhuss) [inline core/src/main/java/io/fabric8/maven/core/util/kubernetes/KindFilenameMapperUtil.java]: We already store some meta data in the classpath, but in `META-INF/fabric8` (like the `enrichers-default`). Could we switch the location to `META-INF/fabric8`, too ? And maybe use the `kind-fileindicator-mapping-default.adoc` as name ? (to adhere to the current convention to add `-default` towards the end).

File: core/src/main/resources/fabric8/default-kind-filename-mapping.adoc (hunk 1 above)
Comment (lordofthejars) [reply core/src/main/resources/fabric8/default-kind-filename-mapping.adoc]: In the original solution, https://github.com/fabric8io/fabric8-maven-plugin/blob/5e52eded75357bc8df3ac4ce93e931bd39c75e23/core/src/main/java/io/fabric8/maven/core/util/kubernetes/KubernetesResourceUtil.java#L198-L244 first was the "short" version and second the "long" version so what you suggest here is to leave the header and change the columns, ok, it should not be hard at all.

About the name, how about `filename-suffix`?

Comment (rhuss) [reply core/src/main/resources/fabric8/default-kind-filename-mapping.adoc]: > About the name, how about `filename-suffix`?

Sound goods. We could also call it 'filename type' (not sure if 'suffix' pronounces the location of this type a bit too much). But never mind, we can change that easily later when we found something better.

File: core/src/main/java/io/fabric8/maven/core/util/AsciiDocParser.java [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+
+        if (column != null && column.startsWith(END_TABLE)) {
+            return END_TABLE;
+        }
+
+        if(column == null || !column.isEmpty()) {
+            throw new IllegalArgumentException(String.format("Trying to read an empty line for end of row, but content %s was found or EOF", column));
+        }
+
+        return column.trim();
```
Comment (rhuss) [inline core/src/main/java/io/fabric8/maven/core/util/AsciiDocParser.java]: I guess you could return a plain "" directly, otherwise the line above would have thrown an exception anyway.

File: core/src/main/java/io/fabric8/maven/core/util/kubernetes/KindFilenameMapperUtil.java [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+        } catch (IOException e) {
+            throw new IllegalStateException(e);
+        }
+    }
+
+    private static InputStream loadContent(String location) {
+        InputStream resourceAsStream = KindFilenameMapperUtil.class.getResourceAsStream(location);
+
+        if (resourceAsStream == null) {
+            throw new IllegalArgumentException(String.format("%s cannotbe foind in classpath", location));
```
Comment (rhuss) [inline core/src/main/java/io/fabric8/maven/core/util/kubernetes/KindFilenameMapperUtil.java]: typo: "cannotbe foind"

File: core/src/main/java/io/fabric8/maven/core/util/AsciiDocParser.java [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+            throw new IllegalArgumentException("Trying to read a column but white line or EOF was found.");
+        }
+
+        int separator;
+        if ((separator = column.indexOf("|")) < 0) {
+            throw new IllegalArgumentException(String.format("Expected the initial of a column with (|) but %s found.", column));
+        }
+
+        return column.trim().substring(separator + 1)
+            .replaceAll("[`_*]", "");
```
Comment (rhuss) [inline core/src/main/java/io/fabric8/maven/core/util/AsciiDocParser.java]: I would put another trim after the `replaceAll` to catch also things like `* dc*, ..`. Not very likely but I think would be a quick win.

File: core/src/test/java/io/fabric8/maven/core/util/AsciiDocParserTest.java [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.catchThrowable;
+
+public class AsciiDocParserTest {
+
+    private static final String VALID_TABLE = "cols=2*,options=\"header\"]" + System.lineSeparator()
+        + "|===" + System.lineSeparator()
+        + "|Filename" + System.lineSeparator()
+        + "|Kind" + System.lineSeparator()
```
Comment (rhuss) [inline core/src/test/java/io/fabric8/maven/core/util/AsciiDocParserTest.java]: would exchange the headers here.

File: core/src/test/java/io/fabric8/maven/core/util/AsciiDocParserTest.java (hunk 2 above)
Comment (rhuss) [reply core/src/test/java/io/fabric8/maven/core/util/AsciiDocParserTest.java]: obsolete as we decided to go for Kind, Filename.

File: core/src/main/java/io/fabric8/maven/core/util/kubernetes/KubernetesResourceUtil.java [hunk 9]
```diff
@@ -182,64 +182,29 @@ public boolean accept(File dir, String name) {
...
+
+            final List<String> filenameTypes = entry.getValue();
+            final String kind = entry.getKey();
+            for (String filenameType : filenameTypes) {
+                FILENAME_TO_KIND_MAPPER.put(filenameType, kind);
+            }
+
+            // In previous version, last one wins, so we do the same.
+            KIND_TO_FILENAME_MAPPER.put(kind, filenameTypes.get(filenameTypes.size() - 1));
+
```
Comment (rhuss) [inline core/src/main/java/io/fabric8/maven/core/util/kubernetes/KubernetesResourceUtil.java]: yeah, we should also document this.

File: core/src/main/resources/META-INF/fabric8/kind-fileindicator-mapping-default.adoc [hunk 10]
```diff
@@ -0,0 +1,102 @@
+|===
```
Comment (rhuss) [inline core/src/main/resources/META-INF/fabric8/kind-fileindicator-mapping-default.adoc]: We should add a documentation in the header (as mentioned in a previous comment) and we should add that the canonical name should be the last part of the mapping (which I would say is the full kind in all lowercase).

File: core/src/main/resources/META-INF/fabric8/kind-fileindicator-mapping-default.adoc [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+|Kind |Filename Type
+
+|ConfigMap
+a|`cm`, `configmap`
+
+|CronJob
+a|`cronjob`, `cj`
```
Comment (rhuss) [inline core/src/main/resources/META-INF/fabric8/kind-fileindicator-mapping-default.adoc]: "cronjob" should go last

File: core/src/main/resources/META-INF/fabric8/kind-fileindicator-mapping-default.adoc [hunk 12]
```diff
@@ -0,0 +1,102 @@
...
+
+|ClusterRole
+a|`cr`, `crole`, `clusterrole`
+
+|CustomResourceDefinition
+a|`crd`
```
Comment (rhuss) [inline core/src/main/resources/META-INF/fabric8/kind-fileindicator-mapping-default.adoc]: would add a "customresourcedefinition" although its quite lengthy.

File: core/src/main/resources/META-INF/fabric8/kind-fileindicator-mapping-default.adoc [hunk 13]
```diff
@@ -0,0 +1,102 @@
...
+
+|ClusterRoleBinding
+a|`crb`, `clusterrb`
+
+|Deployment
+a|`deployment`
+
+|ImageStream
+a|`is`
```
Comment (rhuss) [inline core/src/main/resources/META-INF/fabric8/kind-fileindicator-mapping-default.adoc]: \+ "imagestream"

File: core/src/main/resources/META-INF/fabric8/kind-fileindicator-mapping-default.adoc [hunk 14]
```diff
@@ -0,0 +1,102 @@
...
+
+|ImageStreamTag
+a|`istag`
```
Comment (rhuss) [inline core/src/main/resources/META-INF/fabric8/kind-fileindicator-mapping-default.adoc]: + "imagestreamtag"

File: core/src/main/resources/META-INF/fabric8/kind-fileindicator-mapping-default.adoc [hunk 15]
```diff
@@ -0,0 +1,102 @@
...
+
+|Job
+a|`job`
+
+|LimitRange
+a|`lr`, `limitrange`
+
+|Namespace
+a|`ns`, `clusterrb`
```
Comment (rhuss) [inline core/src/main/resources/META-INF/fabric8/kind-fileindicator-mapping-default.adoc]: hmm. not really sure what "clusterrb" should be and why we allow a namespace resoure at all. I know its not part of the PR, I just thinking out loud ;-)

File: core/src/main/resources/META-INF/fabric8/kind-fileindicator-mapping-default.adoc [hunk 16]
```diff
@@ -0,0 +1,102 @@
...
+
+|OAuthClient
+a|`oauthclient`
+
+|PolicyBinding
+a|`pb`
```
Comment (rhuss) [inline core/src/main/resources/META-INF/fabric8/kind-fileindicator-mapping-default.adoc]: + "policybinding"

File: core/src/main/resources/META-INF/fabric8/kind-fileindicator-mapping-default.adoc [hunk 17]
```diff
@@ -0,0 +1,102 @@
...
+
+|PersistentVolume
+a|`pv`
```
Comment (rhuss) [inline core/src/main/resources/META-INF/fabric8/kind-fileindicator-mapping-default.adoc]: + "persistenvolume"

File: core/src/main/resources/META-INF/fabric8/kind-fileindicator-mapping-default.adoc [hunk 18]
```diff
@@ -0,0 +1,102 @@
...
+
+|PersistentVolumeClaim
+a|`pvc`
```
Comment (rhuss) [inline core/src/main/resources/META-INF/fabric8/kind-fileindicator-mapping-default.adoc]: + "persistenvolumeclaim"

File: core/src/main/resources/META-INF/fabric8/kind-fileindicator-mapping-default.adoc [hunk 19]
```diff
@@ -0,0 +1,102 @@
...
+
+|Project
+a|`project`
```
Comment (rhuss) [inline core/src/main/resources/META-INF/fabric8/kind-fileindicator-mapping-default.adoc]: same comment as for "namespace": Should this be really part of an application descriptor? These descriptors are stored along with other in the jar file, but the "project" or "namespace" is clearly depending on the environment where the application is installed.

File: core/src/main/resources/META-INF/fabric8/kind-fileindicator-mapping-default.adoc [hunk 20]
```diff
@@ -0,0 +1,102 @@
...
+
+|ProjectRequest
+a|`pr`
```
Comment (rhuss) [inline core/src/main/resources/META-INF/fabric8/kind-fileindicator-mapping-default.adoc]: + "projectrequest". I stop here, I think you get the idea for the other rows, too ;-)

Comment (rhuss) [reply core/src/main/resources/META-INF/fabric8/kind-fileindicator-mapping-default.adoc]: Again, not sure what a projectrequest is at all.

**Comment by rhuss** [comment]
I squashed it to a single commit so that its easier to cherry pick to the 3.5.x branch (just in case)
